File: SiHyeon/week6/달리기 경주.cpp
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

vector<string> solution(vector<string> players, vector<string> callings) {
    unordered_map<string, int> rank;
    for (int i = 0; i < players.size(); i++)
    {
        rank.insert(make_pair(players[i], i));
    }
    for (auto c : callings)
    {
        //사람 교체
        string player = players[rank[c] - 1];
        players[rank[c] - 1] = c;
        players[rank[c]] = player;
        //순위 교체
        rank[c] -= 1;
        rank[player] += 1;
    }
    return players;
}

File: hyeyoung/week6/달리기경주.java
package hyeyoung.week6;

import java.util.HashMap;

public class 달리기경주 {
}

class Solution1 {
    public String[] solution(String[] players, String[] callings) {
        String[] answer = new String[players.length];

        HashMap<String, Integer> Player = new HashMap<>();
        HashMap<Integer, String> Rank = new HashMap<>();

        for (int i = 0; i < players.length; i++) {
            Player.put(players[i], i);
            Rank.put(i, players[i]);
        }

        for (String calling : callings) {

            int currentRank = Player.get(calling);
            String player = Rank.get(currentRank);

            String frontPlayer = Rank.get(currentRank - 1);

            Player.put(player, currentRank - 1);
            Player.put(frontPlayer, currentRank);

            Rank.put(currentRank - 1, player);
            Rank.put(currentRank, frontPlayer);
        }

        for (int i = 0; i < players.length; i++) {
            answer[i] = Rank.get(i);
        }

        return answer;
    }

}